src/history/hs.c

```diff
@@ -642,6 +642,15 @@ __wt_hs_insert_updates(WT_CURSOR *cursor, WT_BTREE *btree, WT_PAGE *page, WT_MUL
         for (; upd != NULL; upd = upd->next) {
             if (upd->txnid == WT_TXN_ABORTED)
                 continue;
+
+            /*
+             * We may end up with multiple consecutive tombstones on the update chain if running in
```

Comment:
Is this right. Don't we need the start time pair from the first tombstone and the stop time pair from the second one?

Reply:
Why? We don't write tombstones to the history store. It is only used to set the timestamp of the previous update. If we see multiple tombstones, the stop time pair of the updating they are deleting should be set to the oldest tombstone's.

Reply:
One thing I'm worried is that the tombstones can be committed out of order if we run with READ_UNCOMMITTED and we may end up like (T@10 -> T@20 -> U@5). In that case, we should set the stop timestamp to 10 not 20. I'm not sure whether that is possible?

Reply:
Things become more complicated if we also need to consider txnid. I think we need to set the stop transaction id to the one which is first committed. But in READ_UNCOMMITTED case, we don't know which one is committed first.

Reply:
These seem like good questions - I think a Python test is the right place to make sure we understand the possible scenarios and handle them correctly.

Reply:
@agorrod I'm not sure if it is possible to write a python test to reproduce this. It only happens in a case when two sessions try to delete the same key concurrently. I don't think in python we can reliably replay the same execution sequence.

Reply:
If the out of order timestamp is valid, it can also happen in other types of updates. If we move them to history store, the search results will be wrong. We are facing a general problem here not only just for tombstones.

Reply:
For the out of order timestamp issue, I created WT-5869.

Reply:
> It only happens in a case when two sessions try to delete the same key concurrently

I don't believe it needs to be concurrent - I believe they just need to be interleaved. I think this test could be recreated using multiple sessions in a single thread.

Reply:
@agorrod The case where we put multiple tombstones on the update is a race condition.
The two sessions have to search the key concurrently and both thinks the key is valid. Otherwise, if we first call remove in session 1 and put a tombstone on the update chain and then call remove in session 2, session 2 will see the tombstone added by session 1 and return as it thinks the key is invalid.